kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java

```diff
@@ -72,7 +72,7 @@ public String channelDescriptor() {
      */
     @Override
     public ByteBufferOutputStream createByteBufferOutputStream(int initialCapacity) {
-        final ByteBuf buffer = channelContext.alloc().heapBuffer(initialCapacity);
+        final ByteBuf buffer = channelContext.alloc().ioBuffer(initialCapacity);
```

Comment:
heap buffer(s) in Netty are still pooled, that's why the GC wasn't under pressure - but try performing an heap dump after `::heapBuffer`  and you'll see what it means

@SamBarker

Reply:
Thanks for the clarification @franz1981 I'll spend some more time poking at it, but I think it's slightly orthogonal to the point about our current benchmarking practices not really demonstrating anything useful. As the variation between runs is so great we wouldn't be able to spot the level of perf impact this sort of change implies (though I admit that I didn't make that point very clear).

Reply:
Re the benchmarking practive we can open a separate thread, but in short:
- kafka is highly dependent by OS page cache behaviour (and disk behaviour, in some regard)
- kafka clients are't great both in scalability and troughput-wise

These 2 factors running on the same laptop, without cleanly separating resources, will just produce noisy data.
In theory, kafka should be abstracted away/replaced by something more stable itself (if running on a laptop, clearly) - and isolated, while the client(s)...meh; monitor when they run out of buffer(s) and/or get stalled (something that I've myself modified on OpenMessaging benchmark, to be sure is detected - but clients as load gen producers doesn't have OOTB.